### ISSUE TITLE
Redesign filesystem API

### ISSUE THREAD

**Issue by skidding** [react-cosmos/react-cosmos#440]
### What's up?

Now that we [made the proxy API great again](https://github.com/react-cosmos/react-cosmos/pull/422), it's time to target our next victim. 😈

The README starts with this proposition:
> Cosmos **scans your project for components** and enables you to...

Indeed Cosmos scans your project for components, but the API is terrible.  Namely, `componentPaths`, `fixturePaths`, `getComponentName` are rigid and confusing. These options work if A. your project structure is perfectly organized or B. you know how Cosmos works inside out. Not a #PainlessOnboarding!

### Mkay, tell me more...

I've worked with several large codebases lately and I've reached the conclusion that the filesystem API should primarily focus on giving users **full control** over which components to add to Cosmos where to import them from. This is not a matter of improving documentation, but making the API obvious and flexible in the first place.

> Note: My plan is to support the existing options indefinitely and add complementary ones on top. This only affects an isolated part of the project ([react-cosmos-voyager](https://github.com/react-cosmos/react-cosmos/tree/master/packages/react-cosmos-voyager)) so it won't add much maintenance cost.

The new API is meant to accommodate any level of control, from manually targeting and naming fixtures to letting Cosmos _guess_ where your components are.

## Level 0: *Control freak*

Manually target component and fixture files. Supports any file structure and allows manually naming components and fixtures.

```js
// cosmos.config.js
{
  components: {
    Foo: './src/components/Foo',
    Bar: './src/components/Foo'
  },
  fixtures: {
    Foo: {
      one: './src/components/Foo/__fixtures__/one',
      two: './src/components/Foo/__fixtures__/two'
    },
    Bar: {
      one: './src/components/Bar/__fixtures__/one',
    }
  }
}
```

Users with intricate codebases can automate on top of this by using `glob` and similar tools.

## Level 1: *I did it my way*

Manually target component files, place fixtures by convention. Supports any component file structure and allows manually naming components.

```js
// cosmos.config.js
{
  components: {
    Foo: './src/components/Foo',
    Bar: './src/components/Foo'
  }
}
```

This will work:

```
components/Foo/index.js
components/Foo/__fixtures__/one.js
components/Foo/__fixtures__/two.js
```

This will also work:

```
components/Foo.js
components/__fixtures__/Foo/one.js
components/__fixtures__/Foo/two.js
```

A nice feature of Level 0 and 1 is the ability to **whitelist which components to load**, instead of adding tons of ignore rules.

> Level 0 and 1 can be implemented and shipped alone. The following levels are here to create a big picture and should be implemented separately to avoid monster PRs.

## Level 2: *Let it be*

Manually target component directories, place components and fixtures by convention. Supports multiple component directories (eg. *components* and *containers*).

```js
// cosmos.config.js
{
  componentDirs: ['./src/components']
}
```

This works similar to the existing `componentPaths`, but should do fewer things. No more `fixturePaths` or `getComponentName`. There is Level 0 and Level 1 for that level of customization.

## Level 3: *How did you do that?*

```js
// cosmos.config.js
// *crickets*
```

Try our best at detecting the component dirs based on popular patterns. Searching for `./components`, `./src/components`, etc. If it doesn't work out we ask the user to define `componentDirs` and go back to Level 2.

Notes:
- [fixturesDir](https://github.com/react-cosmos/react-cosmos/blob/ed0f3152197a3e78757d6b60427c71e1edc33905/packages/react-cosmos-config/src/index.js#L8) will continue to work for the new API, so the `__fixtures__` convention can be customized
- I will not be working on this due to other priorities, so anyone can pick it up
- v2 will be released without this, but it's meant to be backwards compatible anyway

### What do you think?

@flaviusone @NiGhTTraX @ovidiubute @tkjone @xavcz @amrdraz @jlc467 @maciej-ka This is a draft so I'd love to hear your feedback!

**Comment by amrdraz** [comment]
Regarding Level 3

Can't we detect component by scanning all files and folders starting the root for files that look like they are components

example files with JSX syntax using esprima
or files that import React
initially I was thinking files that return a component but I don't think we should go beyond static analysis.
so maybe use babel to fish out promising abstract syntax trees or something similar
[inspired by reading this](https://medium.com/@kentcdodds/how-writing-custom-babel-and-eslint-plugins-can-increase-your-productivity-and-improve-user-fd6dd8076e26)

maybe I'm over engineering this, a regex that notices that React was imported/required would probably suffice 😅 

certainly a .jsx file is a good hint as well

**Comment by jlc467** [comment]
Level 3 is nice, but #389 is lower hanging fruit IMO. Adding component directories to a config file is simple relative to creating fixtures so cosmos can render them.

Level 3 and #389 done together gives much improved first time experience:

1. yarn add --dev react-cosmos-webpack (this is kind of confusing because the project is react-cosmos)
2. Add  "cosmos": "NODE_ENV=development cosmos" to scripts (weird step but necessary)
3. yarn cosmos
4. Cosmos playground will auto detect component paths with level 3 and present user with a selectable list of components to create fixtures for with #389. 

No config files, no learning new concepts (i.e. fixtures) before seeing the playground in action.

That said, I think #440 needs to be done before #389 since #389 uses code from `react-cosmos-voyager` whereas #440 will use nothing from #389.

Personally, I plan to use level 2, which is basically what i'm using today. Can't speak much for level 0-1 -- never had a need to target so selectively.

**Comment by maxsalven** [comment]
@skidding awesome!

Any chance of nested fixtures in Level 0? And collapsable folders (the open/closed state of which is persisted across sessions) in the component browser would be awesome :).

I've currently got a component with about 50 fixtures, which I'd love to subcategorise.

Would also love to be allowed to add a description to my fixtures which shows up below the fixture name.

Further dream: "Display all" next to folder name, and it shows me all my fixtures in that folder on the screen at once :)

```
{
  components: {
    Foo: './src/components/Foo',
    Bar: './src/components/Foo'
  },
  fixtures: {
    Foo: {
      one: './src/components/Foo/__fixtures__/one',
      two: './src/components/Foo/__fixtures__/two'
    },
    Bar: {
      green: {
         big: './src/components/Bar/__fixtures__/green-big',
         small: './src/components/Bar/__fixtures__/green-small',
      },
      blue: {
         big: './src/components/Bar/__fixtures__/blue-big',
         small: './src/components/Bar/__fixtures__/blue-small',
    },
  }
}
```

**Comment by flaviusone** [comment]
Love level 0! Makes everything simple and easy to understand. And it can also be used for a feature that lets you extend that config from the Cosmos UI itself (and save it on disk). That would imply having the config config in a separate file.

Level 1 and 2 might be the thing that makes configuring cosmos hard to understand for new users as they have to setup their files by "a convention" and let cosmos do a lot of magic. Maybe something like Jest's [testMatch](https://facebook.github.io/jest/docs/en/configuration.html#testmatch-array-string) or [testRegex](https://facebook.github.io/jest/docs/en/configuration.html#testregex-string) might be enough (with a default glob/regex set by "convention").

@amrdraz's ideas for Level 3 sound good and you can model it as an experimental cli command that generates a config described in Level 0 (much easier to implement).

**Comment by jlc467** [comment]
I like the idea of config generation along with fixture generation. So far though, I seem to be the only one that finds creating fixtures tedious while creating a cosmos config a breeze. I think both have value, but a CLI command to generate a cosmos config seems like helping someone to cross halfway across a busy street. You get them started like hey this is easy, but leave them stranded with components they can't render because they don't have fixtures yet. That is exaggerated because I have noticed the playground does a great job of pointing to the readme which explains fixtures.

**Comment by skidding** [comment]
What a crowd! I suspected this was of interest.

> Can't we detect component by scanning all files and folders starting the root for files that look like they are components

@amrdraz Yup, there's potential here. Wouldn't rely on .jsx files since it's flaky, and you can have .jsx fixture as well. But with static analysis we should be able detect files with React components exported. 👍

> So far though, I seem to be the only one that finds creating fixtures tedious while creating a cosmos config a breeze

@jlc467 I'm in the same boat with you, but we might be a minority. That's not to say that Cosmos should try to please everyone when constraints have a positive impact. To be clear, I wouldn't allow Level 0 to disable the ability to generate fixtures automatically. In fact, Level 0 is the actual data structure we use internally, as the result of file discovery. It's just a matter of cleaning up our architecture to make it available for the user to deliver this data structure by hand. Worse case scenario, automatic fixture generation can only work for configs using Level 1 and above, but Level 0 setup and fixture generation features are complementary.

> No config files, no learning new concepts (i.e. fixtures) before seeing the playground in action.

What if we could merry Level 0 with this?

Imagine cosmos.config.js being as verbose as possible, being able to customize as much as you can, but in most cases never having to change it manually. As you say, we can have auto detect and auto generate fixtures via the Playground UI, but then we can store it in the config so the user sees what happened in the back. Full transparency and customizability, with a UI layer on top. What do you think?

> Any chance of nested fixtures in Level 0? And collapsable folders (the open/closed state of which is persisted across sessions) in the component browser would be awesome :).

@maxsalven Nested fixtures sounds good. Definitely worth keeping in mind. It will work great together with https://github.com/react-cosmos/react-cosmos/issues/251#issuecomment-327676696

> I've currently got a component with about 50 fixtures

😍 

> Would also love to be allowed to add a description to my fixtures which shows up below the fixture name.

I've also been thinking about `fixture.description`, but my thinking was to show it above the Loader in a "meta" header, with possibly other data/actions in the future. Separate thread in any case.

> Makes everything simple and easy to understand. And it can also be used for a feature that lets you extend that config from the Cosmos UI itself (and save it on disk). That would imply having the config config in a separate file.

@flaviusone now I realize you were saying what I mentioned the @jlc467 above. Love the idea of an explicit config that is generated by the UI (or scripts at first, then gradually to build UI around it).

**Comment by jlc467** [comment]
I'm genuinely curious why people want the granularity of level 0. I think some example user stories would help me. Seems time consuming, high maintenance. Sometimes I want to hide certain components that don't have fixtures, so from that perspective I can see the benefit if it happened often enough.

>  As you say, we can have auto detect and auto generate fixtures via the Playground UI, but then we can store it in the config so the user sees what happened in the back. Full transparency and customizability, with a UI layer on top. What do you think?

Now level 3 storing as level 0 is a different story. Deleting components from the config that you inadvertently added through the UI is often easier than deleting through UI. The UI would handle the heavy load of adding the components paths to the config, ideally with a chance to sub-select from the detected components. If we are detecting components, we can detect any existing fixtures too, and if not found, give the option to scaffold them. I like this idea the best so far.

**Comment by tkjone** [comment]
**Background:**

I am working on a Large enterprise react/redux application.  The project is greenfield, so we have it easier than most.  The project also requires us to build out a custom ui-library for internal use across a min of 5 other products.  We are following open source standards as best we can.

With this in mind, my preference is **level 0**

**Pros:**

* Supports all project structures
* Clear expectations about what appears in the playground.  This is great when working with a large applications when you want to control exactly what appears in the playground as large applications, generally speaking, have more components and not all of them are expected to be rendered independently.  This is especially important for our app because its not just developers who are looking at this, it is business and design.  Making the folder structure easy to understand goes a long way in getting continued business support and assisting design.
* No need to keep expanding my `ignore` list.  Consider; My `ignore` list is growing all the time as we develop components that are more containers than anything else.  This makes me wish I could just specify the exact structure I want.
* Avoids _magic_.  If you are new to `react-cosmos`, and maybe even new to React, and just want to play with `cosmos` in a `react-create-app` the current suggested `cosmos.config.js` won't actually work without the developer adjusting the project structure of their `react-create-app` (you have to add a `components` dir and add some components etc).  This is now a min of two extra steps to start using cosmos.  In the case I just mentioned, this is a loss because all the components provided out-of-the-box by `react-create-app` can be easily rendered inside of `cosmos` without proxies or any additional setup.  It would be helpful if people new to `cosmos` understood, from the beginning, how to get a component inside of `react-cosmos` which is what, I believe, would happen with `level 0`.  Please note I have left out the intermediate and advanced developers because my assumption (right or wrong) is that they are capable of figuring out how to make cosmos work and the current docs that are here will be enough for them to work with.  

**Cons**

* Boilerplate.  I don't actually think this is that much of an issue because there are other libraries, like Redux, which are interacted with all the time and people are generally fine with this.  Yes, people build out boilerplate reducing libraries, but even without libraries, its not so bad.

There was a comment about maintenance.  I am curious what this is in reference to.  Would you be able to expand @jlc467?

**Comment by jlc467** [comment]
@tkjone Thanks for the writeup. By maintenance I mean -- what if you move a component to another folder, rename a folder, etc -- my understanding is this would all require updating level 0 config. 

These components on your ignore list -- I'm assuming you don't create fixtures for them -- what if components without fixtures were hidden from view? Would you still want level 0 control?

I agree about `create-react-app`, it is an annoying step that could be replaced with something that contributes to learning cosmos. So how would a level 0 `create-react-app` config look..like this?
```
{
  components: {
    App: './src/App'
  },
  fixtures: {
    App: {
      one: './src/__fixtures__/App'
    }
  }
}
```

It actually looks nice with just one component.. trying to imagine 200.

I can see a ui library that is shared across multiple teams wanting the explicit level 0 config for the sake of self-documentation and reliability (no surprise components added accidentally). Also, if I imagine my team growing in size, I could start to see level 0 making sense. More developers adding components, some of which aren't meant for cosmos -- could create clutter/confusion -- level 0 would ensure that components that are in cosmos are meant to be there.

**Comment by skidding** [comment]
> Now level 3 storing as level 0 is a different story

We're on to something here...

I see this discussion converging into a broader, more exciting idea than the initial draft. I'll try to sum up my ideal outcome based on the ideas shared so far:

- Full control over file paths in config (more below regarding Level 0 vs Level 1)
- File paths can be detected and automatically added to config via UI onboarding
- Next step: New fixtures can also be generated via UI and their paths will be automatically added to config based on the same logic (this is out of scope here, but needs to be taken into account)

This means a **single way to config paths** with complementary UI flows to automate the entire process. Thus, when fixture detection/generation falls short user can go into config and have complete control.

A big benefit of this is: **Component whitelisting**, something that it's painful to do now.

That said, I don't see any value in whitelisting fixtures. Components can exist for various purposes, but we only add files to fixture dirs if we want them to load inside Cosmos. So I think **Level 1** would be a cleaner alternative where we don't have hundreds of needless file paths in our config.

How does everybody feel about this summary? Did I miss an important part?

**Comment by skidding** [comment]
> In the case I just mentioned, this is a loss because all the components provided out-of-the-box by react-create-app can be easily rendered inside of cosmos without proxies or any additional setup

@tkjone good point, I didn't realize this.

**Comment by skidding** [comment]
OK, I'm going to build on my previous comment to start speccing this feature and get it closer to reality.

It seems like we're all psyched on config generation, and I've created a dedicated thread to start building a strategy in that direction #452. For the purpose of this thread, however, I say Level 2 and 3 go away. Until config generation becomes a thing, we can continue to rely on `componentPaths` if we want to avoid verbose path configuration.

**The chief issue here is that it is currently impossible for many people to configure Cosmos. And if not impossible, it's at least annoying.** So let's take another look at **Level 0** and **Level 1** and try to nail this part first.

## Level 0

This is meant to be **Plan B** and will not be the recommended way to specify file paths. Why isn't this Plan A?
- Because whitelisting fixtures is an improbable desire
- Because configs will get huge and repetitive if we specify each fixture path
- Because we don't benefit by new fixture files being added on the fly via hot reloading (instead, you'd need to restart the Cosmos process to see the new fixture)

So why are we still doing it then?
- Because we want to ensure *anything* is possible. We've already made wrong assumptions about file paths in the past and don't want to have to return to this problem again
- Because it doesn't cost us much. The goal is to accept a data structure for fixture paths that we also resolve to in Level 1. It would be a pity to not let the user skip some magic if they really need to. In other words, the pseudo-code looks like this

  ```js
    const components = config.components;
    const fixtures = config.fixtures || findFixtures(components);
  ```

## Level 1

This is meant to be **Plan A** until we come up with config generation magic (again: `componentsPaths` will continue to work for existing users).

So, we already know it will look like this:

```js
// cosmos.config.js
{
  components: {
    Foo: './src/components/Foo',
    Bar: './src/components/Foo'
  }
}
```

And we've established that this allows us to:
- Structure component files however we want
- Whitelist components
- Gradually adopt Cosmos (this is a big one, the current solution expects you to prepare all your components for Cosmos as soon as you start!)

So the big question remaining is: **What about fixtures?**

It would be nice to offer some flexibility without having to resort to Level 0. For example, I always use `__tests__` with Jest, so I find `__fixtures__` natural, but over time I've come to see that `.test.js` is more popular, so maybe we support `.fixture.js` as well.

Moreover, building on what @flaviusone was saying earlier, it would be nice to follow the footsteps of Jest and support custom patterns for fixture paths. To be fair this is a more difficult problem since in Jest you only need to detect tests, whereas here we want to detect fixtures corresponding to a component. That said, I think it is still doable by using vars like `<componentDir>` and `<componentFilename>` in paths.

```js
fixturePathPatterns: [
  '<componentDir>/__fixtures__/<componentFilename>/**/*.{js,jsx,json}',
  '<componentDir>/__fixtures__/**/*.{js,jsx,json}',
  '<componentDir>/*.fixture.{js,jsx,json}'
]
```

Let's take a closer look at each of these:

#### Flat `__fixtures__`

```
<componentDir>/__fixtures__/<componentFilename>/**/*.{js,jsx,json}
```
```
components/Button.js
components/Dropdown.js
components/__fixtures__/Button/default.js
components/__fixtures__/Dropdown/default.js
```

#### Nested `__fixtures__`

```
<componentDir>/__fixtures__/**/*.{js,jsx,json}
```
```
components/Button/index.js
components/Button/__fixtures__/default.js
components/Dropdown/index.js
components/Dropdown/__fixtures__/default.js
```

#### Nested `*.fixture.js`

```
<componentDir>/*.fixture.{js,jsx,json}
```
```
components/Button/index.js
components/Button/default.fixture.js
components/Dropdown/index.js
components/Dropdown/default.fixture.js
```

But there's one unsolved problem with customizable `fixturePathPatterns`: **How do we infer the fixture name?** At the moment I only see two solutions:
1. Don't allow nested fixtures (like @maxsalven proposed), and the fixture name will be the fixture file name
2. Don't make `fixturePathPatterns` customizable, in which case we can hardcode these 3 patterns and infer differently per each (similar to the existing solution)

Not very happy with either, but I'm leaning towards the 2nd compromise. In other words, embracing convention over configuration and enabling nested fixture files (which I also think is needed).

What does everybody think? Maybe there's a better idea or insight that I'm missing.

**Comment by maxsalven** [comment]
I'd much prefer to name the fixture within the fixture data. I never really understood why I had to be limited to the hard to read formatting restraints of a filename. I want to see my fixture in the UI as "Blue outline with gold flair" and not "blueOutlineWithGoldFlair" or "blue-outline-with-gold-flair".

As discussed elsewhere, I'd love to be able to add a description as well.

I may be in the minority here, but I'd prefer to have all my fixtures for a component be in one file, just like I can with jest tests. In fact, I'd love to mirror jest as much as possible.

```
components/Button/index.js
components/Button/index.fixtures.js
```

If there is no default export, we consider all the exports to be fixtures:
```
const baseProps = { something: true }

export const fixture1 = {
  name: "Defaults", // optional, can use 'fixture1' as name if no name attribute provided.
  props: { ...baseProps },
}

export const fixture2 = {
  name: "Blue Outline with Gold Flair",
  props: {
    ...baseProps,
    color: "blue",
    flair: "gold",
  },
}
```

If you want folders, then you can create a default export:
```
export default { fixtureFolder1: { fixture1 }, fixtureFolder2: { fixture 2 } }
```

I'd also think 'zero setup' is doable using this method, just like in jest. Simply watch for all `*.fixtures*` files, and boom, that's what you show in `cosmos`.

It's basically zero-config, similar to what you discussed in your Death of Config issue. Install `react-cosmos`, specify your webpack config (or autodetect for common use cases), create a fixture (either manually or via a new UI). That's all there is to it. Want another component in cosmos? Just create a fixture for it. No need for component paths config, or whitelists, or ignores, or fixture paths, etc. **If a valid fixture exists, it's in cosmos.**

I'd go one step further and get rid of folder/filename matching completely. It's just a source of headaches. Let's import the component we want inside the fixture, just like in Jest, then it doesn't matter where I put my fixture files, and I can move them around without having to change anything, anywhere.

```
import MyComponent from "src/somewhere/MyComponent"

export const default = {
  component: MyComponent,
  props: { },
}
```

I appreciate this is a bit of a departure from the existing system, but I really think it's far superior. The fixture file 100% defines the fixture now (initialState in a redux reducer, notwithstanding). There is almost zero room for user confusion or error. You can also easily generate these fixtures from the UI.

**Comment by skidding** [comment]
Thanks @maxsalven for the radical ideas! Much needed.

It's funny but when we started off years ago the fixture format used to contain `component` as well. I like the idea of standalone fixtures: a single object that can be used to render a component in a given state, as opposed to a (component, fixture) pair as it is now. One concern is the redundant component imports between more fixtures for the same component, but I think it's worth it if we gain a lot on the usability front. Still, this solution raises some big questions:

- More fixtures in the same file sounds like a great option, but is it a good idea to support it and no longer support one fixture per file? If we're thinking of the Jest metaphor, you can have as many test files as you want for the same component. Additionally, the big difference in Cosmos is we also have a hierarchy of components & fixtures (see #448). We already use a hierarchy to organize our code: the file system, so wouldn't be a pity to create a secondary hierarchy instead of reusing it? I find that part of what makes PHP and Next.js successful is their (re)use of the file system.
- If the fixtures point to components, how are we going to infer the component name for the UI? Not saying anyone should, but it would even be possible to inline components in fixture files.

> You can also easily generate these fixtures from the UI.

I get most of the benefits of this approach, but how does it make it easier to generate fixtures from the UI? I find that the opposite is true: It's easier to generate files when there is convention.

**Comment by skidding** [comment]
@maxsalven I made a playground to experiment with the idea of detecting fixture files only and referencing components in fixtures. **Check this out: https://github.com/react-cosmos/fs-playground.**

I added types and some tests to make it clear, but you can also see the fixture files laid out as randomly as possible. The `'**/__fixture?(s)__/**/*.{js,jsx}', '**/?(*.)fixture?(s).{js,jsx}'` patterns are fully configurable, so in the real world they would be mere defaults. 

Overview:
- Fixtures reference components via `fixture.component`
- Fixture name is read from `fixture.name` and defaults to `default`, `default (1)`, `default (2)`, etc.
- Component name is read from function/Class name (it's easier than I expected)
- Fixture files can be placed anywhere
- Fixture files can export one or more fixtures (didn't use named exports because A. it felt like we're moving away from depending on the file system and B. we should only define the fixture name once via fixture.name, so the default export can be an Array instead)

Unhandled:
- Namespacing. Currently you could name a fixture `namespace/fixture name`, but components should also be able to define their namespace (maybe we can support a static property like `Component.namespace` or `Component.label`)

Wdyt?

**Comment by skidding** [comment]
**Update 🥁:** I have a completed the experimental implementation for a new way of defining fixture files and I'm very excited about the result.

Here are the benefits:
- Fixtures can be placed anywhere, so **no more constraints on the fixture file location**. Fixture files no longer need to be correlated to component files. As @flaviusone suggested, fixtures are now detected using a Jest-like `testMatch` approach. Just like with Jest, I expect almost nobody to change the default value for this option. Per @jlc467's ideas, I think this also makes fixture generation easier since there are no more rigid rules for where to place fixture files.
- **Cosmos can be used without configuration**. Right now the only mandatory config option is `componentPaths`. Unless we need to configure a webpack config path, Cosmos can be made to start working without `cosmos.config.js` because the new file matching works out of the box thru good enough defaults.
- **Fixtures are whitelisted by default**. Because we only show components which have already fixtures, we'll never have to worry about hiding components because we'll never create fixtures for them in the first place. @tkjone I think you'll like this.
- **Fixture files can export one or more fixtures** (for the same or for different components). @maxsalven I know you're a fan.
- **Fixtures can be named** via `fixture.name`, defaulting to "default", "default (1)", "default (2)", etc. if unnamed.
- **Component names are inferred** in a number of steps. First, `Component.displayName` or `Component.name` is checked. Which means **component names can also be manually defined**. But the idea is to detect the name and not ask the user to provide the names unless preferable. So if name isn't detected like this, it will be detected by checking the name of the component import from the fixture file, via static analysis. If the name still couldn't be detected (very improbable), it will default to "Component", "Component (1)", "Component (2)".
- **Component namespace is inferred** from file path. So, like before, if we nest our components in subdirectories, Cosmos will be aware of the relative path to the common dir of all components. Except now it's done without having to specify component paths 🎩. This enables tree view nesting and will work great with the soon to be merged #448 from @maxsalven.
- **Fixture namespace is inferred** from file path. Just like with components, fixture can now be nested in the tree view.
- Both component paths and fixture paths are known and surfaced by the new API, which means we can implement a Playground feature to **click on a component or fixture button and open the corresponding file in the native editor** – A feature I already partially implemented in a late night session with @xavcz but never had time to properly ship cause bigger fish had to be fried first (like this monster issue).

### OK, so what's the catch?

The catch is that the fixture format changes to accommodate all these. A fixture looks exactly the same as before, except with an extra field pointing to the component it belongs to:

```js
import Button from '../Button';

export default {
   component: Button
   props: {}
}
```

That's it.

Still, this might be scary for people with dozens or hundreds of existing fixtures. To combat this I intend to **keep Cosmos backwards compatible.** Instead of refactoring react-cosmos-voyager, I'm thinking of doing something as blunt as creating a new package called `react-cosmos-voyager2` for the new fs logic. I admit I was partially inspired by [this talk](https://www.youtube.com/watch?v=oyLBGkS5ICk0). If the config uses any of the old options (`componentPaths`, `fixturePaths` and `getComponentName`) we use react-cosmos-voyager, otherwise react-cosmos-voyager2. The connecting modules need some extra work to map the old fixtures to the new format, but in theory it's just putting `{ component: Component, ...oldStyleFixture }`. And if anyone wants to create a codemod to help with the transition, even better. I'm also curious to look into that, but I don't want to make it a requirement and break compatibility.

There's also an extra benefit to writing fixtures like this: The Cosmos loader becomes `f(fixture)` instead of `f(component, fixture)` which I think is more elegant as a general abstraction. Especially when using fixtures outside the playground UI, to mock component states in tests.

### File system and AST

I would like to clarify something. A main part of this change is to make fixtures more flexible by decoupling them from the file system, which I think is a good idea. But then I also made use of the file system (through static wizardry) to detect file paths, mainly to use them for generating namespaces which can be nicely turned into a deeply nested tree view mirroring the file system. This might sound contradictory. The reasoning for this is that fixtures are now fundamentally decoupled from the file system, but through the use of **complementary** file system processing we decorate them with extra properties which are 100% optional–and degrade gracefully if they fail in the process. So if the file system logic fails the components won't have a namespace and will be rendered in a flat hierarchy, as before. @amrdraz you might like the AST part.

**All the code is here: https://github.com/react-cosmos/fs-playground**

It started out as an experiment but the code is now typed and has a shitload of tests for every scenario I could think of. It still needs some tuning, but for the most part it could be copy pasted into the monorepo and referenced as a separate module. It would be the first module with Flow types!

Here are the main types to get a picture of the output generated:

```js
type Fixture = {
  name: string,
  namespace: string,
  filePath: string
};

type Component = {
  name: string,
  namespace: string,
  filePath: string | null, // It can be null because detection might fail
  type: ComponentType<*>,
  fixtures: Array<Fixture>
};
```

The root level API will return a list of Component objects, which contain Fixture objects. Feel free to browse [the repo](https://github.com/react-cosmos/fs-playground) to see how the test cases or implementation looks.

## Feedback, please!

While I'm very excited, I also recognize this a major change and want to make sure it's the done right.

This thread along with other feedback I received have revealed a lot and I've come to believe that the current state of file management in Cosmos is the primary obstacle in becoming a more accessible tool which can be of use for many more people. This is why I put so much thought and effort into it.

So, if anybody from this thread can find the time, I would be very grateful if you could let me know:
- What do you think?
- Am I missing something important? (I spent a lot of time with this so could be blind)
- Could this be improved even more?

Thanks! ⚔️

**Comment by amrdraz** [comment]
I'm loving the AST part 😍 
will using require resolve account for aliases? 
for instance I use in one of my projects babel-module-resolver (using next.js)
in another I use webapack's alias

mainly to refer to the root of the project with example `~/components/x`

**Comment by skidding** [comment]
@amrdraz I don't know if it path detection can work with webpack's aliases, but it should be possible to make it work with babel-plugin-module-resolver, which seems to have an [API](https://github.com/tleunen/babel-plugin-module-resolver#for-plugin-authors) for this. It will be nice to see compatibility grow for this in time. For example my POC works with ES modules, but not CommonJS. I left a TODO for that, which can be implemented and released later to broaden support. The dev experience is cool because the tests work with real files so creating new scenarios is straightforward.

The nice thing about it is that if the path detection fails:
- Core functionality still works, except `component.filePath` will be `null` so `component.namespace` will always be `''`.
- Namespaces can still be manually set like `Component.namespace = 'Admin/Header'`.

**Comment by amrdraz** [comment]
I guess I can then just have a decorator that applies the namespace property to the component  using the babel-plugin-module resolver since it feels wrong to have support for as a specific babel plugin built in to this

**Comment by skidding** [comment]
Good point. Or maybe Cosmos can surface an cb option to manually resolve
relative paths (aliases) to absolute paths, and then you can forward that
to babel-plugin-module resolver in cosmos.config. I think this is flexibile
enough to work with other resolvers as well and adds almost 0 complexity to
Cosmos.

On Oct 11, 2017 15:20, "Amr Draz" <notifications@github.com> wrote:

I guess I can then just have a decorator that applies the namespace
property to the component using the babel-plugin-module resolver since it
feels wrong to have support for as a specific babel plugin built in to this

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/react-cosmos/react-cosmos/issues/440#issuecomment-335807010>,
or mute the thread

**Comment by flaviusone** [comment]
Will I still be able to reuse the new format fixtures for my unit tests e.g. shallow rendering?

**Comment by skidding** [comment]
> Will I still be able to reuse the new format fixtures for my unit tests e.g. shallow rendering?

Yes, 100%. The only difference is that the (pseudo) API will be `shallow(loader(fixture))` instead of `shallow(loader(component, fixture))`. So I would say it's even cleaner.

**Comment by skidding** [comment]
@maxsalven asked if he can try this prototype on his codebase, and maybe others are curious as well and want to help test this. It is now possible! 🔥

Three simple steps:
1. `yarn remove react-cosmos-webpack`
2. `yarn add --dev @skidding/react-cosmos-webpack@^2.2.0-1`
3. **Report what happened!**

Here is an example on me doing this with the Flatris repo:
![new-fs-alpha](https://user-images.githubusercontent.com/250750/31518844-cb31dba2-afa0-11e7-91e9-04ea02c3e2b9.gif)

You can see there is a nice warning about incompatible fixtures and they become available as soon as the `component` property is added. If only we had a codemod to automate this with a click, backwards compatibility might not be required after all... 🤔

**Comment by skidding** [comment]
OK, my previous comment was nothing. Check this out: **Automatic fixture upgrade!** 🚨

Three simple steps:

1. `yarn add --dev @skidding/react-cosmos-webpack@^2.2.0-2` (bump from last time)
2. `yarn upgrade-fixtures`
3. **Report what happened!**

Just look at these gifs:
![fixture-autoheal](https://user-images.githubusercontent.com/250750/31568048-b26064ea-b072-11e7-98a9-1eef280a8da4.gif)
![fixture-autoheal2](https://user-images.githubusercontent.com/250750/31568052-b4323b90-b072-11e7-8de4-cb56e6f24eca.gif)

The formatting isn't great yet. It's annoying that [babel-generator](https://github.com/babel/babel/tree/master/packages/babel-generator) removes the comments for example. Tried [recast](https://github.com/benjamn/recast) and wasn't very happy with that one either, but might be a better solution. 

Either way, I'm curious to see how it behaves for others, so please try it out!

**Comment by skidding** [comment]
Moved from babel to [jscodeshift](https://github.com/facebook/jscodeshift) and the output is much better. An extra newline here and there but otherwise it's perfect.

![fixture-autoheal3](https://user-images.githubusercontent.com/250750/31575241-557047bc-b0e2-11e7-926c-a7abb67c2639.gif)

Note that the codemod currently works for ES modules. **Does anybody here have CommonJS fixtures?** cc @flaviusone

**Comment by flaviusone** [comment]
> Does anybody here have CommonJS fixtures? cc @flaviusone

Yes we do and we also use react-cosmos-normalize-props-proxy.

**Comment by jlc467** [comment]
Installed @skidding/react-cosmos-webpack@^2.2.0-2 and gave it a quick spin.. seems to work great.
Couple observations:
- The playground listed a fixture whose component was not in cosmos config `componentPaths`.
- The new fixture format is intuitive/concise -- I like it. Biggest win for me is fixture folder/file structure freedom. I was one of the main things preventing me from writing fixtures.

My next cosmos experiment when I find the energy:Writing a script that you run inside a component folder that detects all the components and creates empty fixtures (using new format) for all of them, either co-located or in a path that is passed.

**Comment by maxsalven** [comment]
I have a somewhat unconventional existing setup, but it works perfectly in the existing release version of `react-cosmos`.

I had essentially created my own 'level 0' config, by having a `src/components/__cosmos__` directory, and a `componentPaths: ["src/components/__cosmos"]` in my config.

If I want to see a component in cosmos, then in my `__cosmos__` directory, I might have something like this:

```
// components/ __cosmos__/ui/Modal/index.js
import Modal from "components/ui/Modal"; // the 'real' component

export default Modal;
```

```
// components/ __cosmos__/ui/Modal/__fixtures__/open.js
import React from "react";

export default {
  props: {
    isOpen: true,
    onDismiss: () => console.log("Closed Modal"),
  },
  children: <div>Children</div>,
};

```

However the codemod has created invalid code:
```
import React from "react";
import ui/Modal from '../index';

export default {
  component: ui/Modal,

  props: {
    isOpen: true,
    onDismiss: () => console.log("Closed Modal"),
  },

  children: <div>Children</div>
};
```

Specifically, it has put a "/" in the import name which is not allowed. The code mod doesn't really apply to my case anyway (as I no longer need to have my workaround system) but just an FYI that there are scenarios where it produces invalid code.

**Comment by skidding** [comment]
Thanks @maxsalven, that was very helpful. Found a fix for the transform which works well locally against a repo with similar nested files. 👍

`3.0.0-alpha.1` coming soon with:
- New fixture API
- Backwards compatibility with old fixtures
- Optional transform for upgrading fixtures (will not work in 100% cases, eg. when using aliases)
- Brand new 🌲 view!

Excited!

**Comment by skidding** [comment]
### The 3.0 alpha is here! 🥁🎺

**Try out this branch by installing `react-cosmos-webpack@next` and `react-cosmos-scripts@next` (for the fixture upgrade codemod).**

Here's a rundown of the Cosmos upgrade in [illustrated-algorithms](https://github.com/skidding/illustrated-algorithms):

```
yarn add --dev react-cosmos-webpack@next react-cosmos-scripts@next
yarn cosmos
```

I get a small notice in the CLI:
```
[Cosmos] Using `componentPaths` config is deprecated. Please consider upgrading.
```

Right off the bat, all my previous fixtures are showing up, nicely listed in the new tree view (courtesy of @maxsalven)
![image](https://user-images.githubusercontent.com/250750/31731350-0c9d0ea2-b435-11e7-987c-da21524cf04d.png)


But if I look closely in the browser console, I get this friendly message:
```
[Cosmos] Successfully read 39 old school fixtures:
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/binary-search/beginning.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/binary-search/comparing.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/binary-search/found.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/binary-search/intro.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/quicksort/intro/blank.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/quicksort/outro/blank.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/quicksort/intro.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/raw-data/first-frame-wip.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-block/blank.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-block/empty.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-block/glow.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-icon/bear.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-icon/cat.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-icon/dog.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-icon/lion.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-icon/no-entry.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-icon/panda.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/emoji-icon/snail.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/label/pivot.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/ill/shared/number-var/max-5.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/menu/binary-search.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/page/bfs.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/page/binary-search.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/page/quicksort.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/playback-controls/finished.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/playback-controls/playing.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/playback-controls/stopped.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/player/binary-search-mid-way.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/player/binary-search.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/player/quicksort.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/source-code/bfs.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/source-code/binary-search.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/source-code/quicksort.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/stack-entry/binary-search-beginning.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/stack-entry/binary-search-finished.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/stack-entry/binary-search-first-assign.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/stack-entry/binary-search-intro.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/stack-entry/binary-search-last-check.js
- /Users/skidding/Work/illustrated-algorithms/components/__fixtures__/stack-entry/binary-search-returning.js
[Cosmos] Upgrade these fixtures by adding the `component` property.
[Cosmos] More details at https://github.com/react-cosmos/react-cosmos/issues/440
```

So I go back to CLI to upgrade fixtures:

```
yarn upgrade-fixtures --config cosmos/cosmos.config.js
```

Aaand, diffs are looking good, fixtures are up to date.

![fixture-autoheal4](https://user-images.githubusercontent.com/250750/31732046-6625ff7c-b437-11e7-8e9b-955635c1dc7f.gif)

Going back to the UI after this change, everything works as before and I don't get any more runtime warnings in the console.

But I still get this CLI notice at compile time:

```
[Cosmos] Using `componentPaths` config is deprecated. Please consider upgrading.
```

So now that the fixtures are upgraded, it's time to get rid of the `componentPaths` and `ignore`, and move to new *auto* config.

![image](https://user-images.githubusercontent.com/250750/31732176-dbbf4fcc-b437-11e7-9752-6c4c45e71b9f.png)

Note that this config is a bit more complicated because the config isn't placed in the project root.

Anyway, I restart the Cosmos server. No more CLI notice. Voila! ✨

![3 0 0-alpha](https://user-images.githubusercontent.com/250750/31732274-3de06d4e-b438-11e7-802c-c7ee10bcc95d.gif)

---

I still have more things to iron out, but please try it out to help detect any hidden issues!

**Comment by jlc467** [comment]
Trying out `react-cosmos-webpack@next`... working great. This is and the new tree view are both huge, amazing steps forward. 👍

> Fixture files can export one or more fixtures (for the same or for different components). 

I might be doing something wrong, but can't get this to go.

Here is my fixture.
```
import Button from '../components/AsyncButton';
import React from 'react';

export const defaultProps = {
  component: Button,
  name: 'Default',
  children: 'Save Item'
};

export const isLoading = {
  component: Button,
  name: 'Loading',
  props: { isLoading: true },
  children: 'Save Item'
};
```
Error:
<img width="458" alt="screen shot 2017-10-18 at 4 41 00 pm" src="https://user-images.githubusercontent.com/6405557/31742174-c8ebd900-b424-11e7-8253-097e8efad2fc.png">

I tried to debug it a little and came up with [this line](https://github.com/skidding/react-cosmos/blob/9613da4a6c2f8ed21d50eb56d93601cf0467fa45/packages/react-cosmos-webpack/src/client/mount.js#L62) which is calling `importModule` that takes a `moduleName` as 2nd parameter to support named es6 exports. However, I can't see that moduleName is being passed so it returns undefined.

**Comment by skidding** [comment]
Hi @jlc467,

Multi fixture files should export an Array as the default export instead of multi exports. Can you try this instead?

```js
import Button from '../components/AsyncButton';
import React from 'react';

export default [
  {
    component: Button,
    name: 'Default',
    children: 'Save Item'
  },
  {
    component: Button,
    name: 'Loading',
    props: { isLoading: true },
    children: 'Save Item'
  }
];
```

**Comment by jlc467** [comment]
@skidding 
https://github.com/skidding/react-cosmos/blob/9613da4a6c2f8ed21d50eb56d93601cf0467fa45/packages/react-cosmos-webpack/src/client/mount.js#L75

**Result when exporting an array like the example you gave:**

At [line 75 ](https://github.com/skidding/react-cosmos/blob/9613da4a6c2f8ed21d50eb56d93601cf0467fa45/packages/react-cosmos-webpack/src/client/mount.js#L75) `componentModule` is undefined because `deprecatedComponentModules` is `{}`. When `componentModule` passed to `importModule`, causes the error below.

![screen shot 2017-10-19 at 4 28 20 pm](https://user-images.githubusercontent.com/6405557/31792650-c498f0e2-b4ea-11e7-93cc-5e6a979492d1.png)

I think the code maybe took a wrong turn because `componentPaths` has to be defined in cosmos config for `deprecatedComponentModules` to be defined. Any ideas?

**Question**: Is it supposed to be read as an old school fixture and then upgraded on the fly? If so, doesn't that involve the AST parser? If so, it seems like there are lot of limitations with the parser.. for example, if the user exports like this: `export default [fixtureOne, fixtureTwo]`,  I was  getting different results than if I exported as in your example. I can give more detail if this is the intended path for multi fixture files.

**Comment by skidding** [comment]
@jlc467 did you remove `componentPaths` from your config before upgrading fixtures? This is the only scenario that makes sense by reading your log. I still don't understand how you got that TypeError since that whole block of code is guarded by try/catch...

> Is it supposed to be read as an old school fixture and then upgraded on the fly? If so, doesn't that involve the AST parser?

This part doesn't involve AST. It just takes the existing fixture and does `{ ...fixture, component }`. But this only works while you still have `componentPaths` in your config. Otherwise you need to upgrade your fixtures as well first, which is done by running the code mod. The code mod is pretty limited indeed, but it should work for old fixtures (eg. no Array exports).

Let me know if I got this right or if you have any other insights.

**Comment by jlc467** [comment]
> I still don't understand how you got that TypeError since that whole block of code is guarded by try/catch...
I logged it to try to understand what is going on.

> did you remove componentPaths from your config before upgrading fixtures?
Never had `componentPaths` defined. Not sure why I should need it, or the code mod for that matter. I'm using newer fixture format, just exporting multiple in a single file. Why does that need to be upgraded?

here is my cosmos config
```js
// cosmos.config.js
module.exports = {
  webpackConfigPath: 'react-scripts/config/webpack.config.dev',
  publicPath: 'public'
};
```

My fixture
```js
import App from './App';
import React from 'react';

export default [
  {
    component: App,
    name: 'Default'
  },
  {
    component: App,
    name: 'Loading'
  }
];
```

**Comment by skidding** [comment]
> Never had componentPaths defined. Not sure why I should need it, or the code mod for that matter. I'm using newer fixture format, just exporting multiple in a single file. Why does that need to be upgraded?

No need. I just assumed you were upgrading old fixtures. My bad. 

OK I think I found the problem and fixed it in https://github.com/react-cosmos/react-cosmos/pull/469/commits/5322d9abbf1894036f2120bc9212fdc30d544aee. Published `react-cosmos-webpack@3.0.0-alpha.5` under the `next` tag. Please try it out and report what happened.

**Comment by maxsalven** [comment]
Latest cosmos alpha worked perfectly for me with my existing fixtures and with the code mod!

Just a small note on the codemod, it doesn't respect `.prettierrc` and would be cool if it did. Perhaps you can check if the file exists, and if so, run prettier before saving the fixture. Not a big deal at all, but if it's an easy win, maybe worth throwing in there.

**Comment by amrdraz** [comment]
for me it doesn't seem to work


first I was getting for configuration

```js
module.exports = {
  componentPaths: ["./components"],
  globalImports: ["static/css/reset.css"],
  ignore: [".test.", /__test__/],
  publicPath: "./static",
 }
```

```
[Cosmos] See you at http://localhost:8989/
[Cosmos] Using `componentPaths` config is deprecated. Please consider upgrading.
webpack built 86ece583b4e300ec97bb in 11410ms
Hash: 86ece583b4e300ec97bb
Version: webpack 3.3.0
Time: 11410ms
  Asset     Size  Chunks  Chunk Names
   0.js   148 kB       0  chunks/react_image_lightbox_65979c8056b2b266b92ff6a8853dd6eb.js
main.js  4.39 MB       1  main
  [70] (webpack)/buildin/module.js 517 bytes {1} [built]
  [88] ./node_modules/babel-runtime/helpers/extends.js 544 bytes {1} [built]
 [199] ./node_modules/querystring-es3/index.js 127 bytes {1} [built]
 [201] ./node_modules/react-cosmos-webpack/lib/client/mount.js 5.34 kB {1} [built]
 [459] multi ./static/css/reset.css (webpack)-hot-middleware/client.js?reload=true ./node_modules/react-cosmos-webpack/lib/client/loader-entry.js 52 bytes {1} [built]
 [460] ./static/css/reset.css 360 bytes {1} [built] [failed] [1 error]
 [461] (webpack)-hot-middleware/client.js?reload=true 6.68 kB {1} [built]
 [464] (webpack)-hot-middleware/node_modules/strip-ansi/index.js 161 bytes {1} [built]
 [466] (webpack)-hot-middleware/client-overlay.js 1.82 kB {1} [built]
 [467] ./node_modules/ansi-html/index.js 4.26 kB {1} [built]
 [471] (webpack)-hot-middleware/process-update.js 3.88 kB {1} [built]
 [472] ./node_modules/react-cosmos-webpack/lib/client/loader-entry.js 394 bytes {1} [built]
 [473] ./node_modules/react-cosmos-webpack/lib/client/react-devtools-hook.js 559 bytes {1} [built]
 [492] ./node_modules/react-cosmos-webpack/lib/client/user-modules.js 54.9 kB {1} [built]
 [886] ./node_modules/react-cosmos-loader/lib/index.js 744 bytes {1} [built]
    + 898 hidden modules

ERROR in ./static/css/reset.css
Module parse failed: /Users/draz/web/cairousel-web/static/css/reset.css Unexpected token (5:37)
You may need an appropriate loader to handle this file type.
| a, abbr, acronym, address, big, cite, code,
| del, dfn, em, img, ins, kbd, q, s, samp,
| small, strike, strong, sub, sup, tt, var,
| b, u, i, center,
| dl, dt, dd, ol, ul, li,
 @ multi ./static/css/reset.css (webpack)-hot-middleware/client.js?reload=true ./node_modules/react-cosmos-webpack/lib/client/loader-entry.js
```

which implied the css loader isn't included which I figured is fine since I can allways wrap my components

commenting out `globalImports` leaving only 

```js
module.exports = {
  componentPaths: ["./components"],
  ignore: [".test.", /__test__/]
}
```

lead to cosmos running but a 404 error when I visited http://localhost:8989

The console stated `GET 404  http://localhost:8989/loader/index.html`

this was on a next.js project

I tried starting a fresh next.js project using create-next-app, got the following in the terminal

```
[Cosmos] Using defaults, no config file found
[Cosmos] No webpack config found, using default config
[Cosmos] Using defaults, no config file found
[Cosmos] See you at http://localhost:8989/
[Cosmos] Using defaults, no config file found
webpack built ccaba5f54325e5aec91f in 1681ms
```
looking at `http://localhost:8989/` resulted in a similar result

<img width="1278" alt="screen shot 2017-10-20 at 23 13 03" src="https://user-images.githubusercontent.com/1708559/31842006-4ad135be-b5ec-11e7-8fa9-1aafeb1ddbc8.png">

so I added 

```js
// cosmos.config.js
module.exports = {
  rootPath: ".",
  publicPath: "static",
  publicUrl: "/static/"
}
```

but still nothing

I sense something is wrong with the default webpack config file?

**Comment by amrdraz** [comment]
also when I tried running 
```
npm run upgrade-fixtures
```

I would get

```
Error: You cannot call "get" on a collection with no paths. Instead, check the "length" property first to verify at least 1 path exists.
```

but I noticed some of the files where getting converted so I walked through them until I noticed that the codemod doesn't seem to work if the fixture is exported using `module.exports`

renamed it to `export default` and everything went well (in case anyone faces that problem)

**Comment by amrdraz** [comment]
I'm realizing that the issue was apparently there for a while and I had not upgraded since `2.0.0-beta.19` so I'll be upgrading my configuration after reading through the release notes and working my way through there

**Comment by skidding** [comment]
Sorry if this was communicated poorly, but the default config has changed a while ago and it doesn't come pre packed with any the loaders it supports. This was because there were too many collisions possible so it's best to let each user install the version of the loaders they way. After all, Cosmos works with webpack 1 to 3. Here is the corresponding README section

![image](https://user-images.githubusercontent.com/250750/31843874-af77e702-b5f5-11e7-9c2c-d746759a39bc.png)

Once you have all the loaders installed, the Cosmos default webpack config will detect and use them automatically. Relevant thread https://github.com/react-cosmos/react-cosmos/issues/407

> lead to cosmos running but a 404 error when I visited http://localhost:8989

> The console stated GET 404 http://localhost:8989/loader/index.html

This is because the above also applies for `html-webpack-plugin`. Make sure to install it if you don't use a custom webpack config.

> but I noticed some of the files where getting converted so I walked through them until I noticed that the codemod doesn't seem to work if the fixture is exported using module.exports

Indeed, if you look at my [comment](https://github.com/react-cosmos/react-cosmos/issues/440#issuecomment-336628485) 6 days ago it says

> Note that the codemod currently works for ES modules. Does anybody here have CommonJS fixtures?

This can be extended later (and anyone can pick this up, AST transforms are pretty straightforward to test and develop), but there was so much to do I tried to save time wherever I could. Plus I assumed there are existing transforms published for transforming ES modules into CJS modules, so upgrading fixtures could be achieved in two steps.

@amrdraz hope this helps. Let me know if you hit any other roadblocks. I appreciate you testing this!

**Comment by amrdraz** [comment]
@skidding I eventually got everything working, thank you for clarifying, everything went smooth for me after, should have noticed the new getting started 😅 .

I'm excited about being able to decorate fixtures now that I'm passing the component myself

regarding the commonjs were there any valid commonJS cases other than `module.exports`?

I noted an empty fixture file which caused the same error for fixture upgrade, though I don't think that was a valid case for a fixture to begin with.

**Comment by amrdraz** [comment]
@skidding just noticed folders that appear to be getting added from nothing in the tree view

I haven't figured out why yet but I observe that given folder strucure

<img width="639" alt="screen shot 2017-10-21 at 02 12 52" src="https://user-images.githubusercontent.com/1708559/31845845-bc4971fc-b605-11e7-86d6-1f825e64fa10.png">

I got an extra folder called NavigationTopBar inserted when it should have been just TopBar

<img width="362" alt="screen shot 2017-10-21 at 02 11 19" src="https://user-images.githubusercontent.com/1708559/31845846-bc9313ca-b605-11e7-9d7f-68dd160e5a8f.png">

I also noticed that in general folders that host components show up as both a folder and a component for all components, but I guess this is just by design.

<img width="604" alt="screen shot 2017-10-21 at 02 25 58" src="https://user-images.githubusercontent.com/1708559/31846096-722a6f34-b607-11e7-80ba-a74b5a36e14e.png">
<img width="341" alt="screen shot 2017-10-21 at 02 27 22" src="https://user-images.githubusercontent.com/1708559/31846097-72642ada-b607-11e7-8294-2e2d6f137deb.png">

folders that are just folders appear once

<img width="764" alt="screen shot 2017-10-21 at 02 31 41" src="https://user-images.githubusercontent.com/1708559/31846159-128613d4-b608-11e7-8dbd-ebd52d8d8c11.png">
<img width="331" alt="screen shot 2017-10-21 at 02 18 53" src="https://user-images.githubusercontent.com/1708559/31846095-71ef35fe-b607-11e7-82da-4d0cc804f81b.png">

**Comment by skidding** [comment]
> I eventually got everything working, thank you for clarifying, everything went smooth for me after, should have noticed the new getting started

Awesome sauce!

> regarding the commonjs were there any valid commonJS cases other than module.exports?

Not that I'm aware of.

>  just noticed folders that appear to be getting added from nothing in the tree view

OK, just released `3.0.0-alpha.7` which fixes the tree nesting excess. Check it out @amrdraz @maxsalven 

![image](https://user-images.githubusercontent.com/250750/31851652-6e89be76-b66a-11e7-8995-a153d6d408d8.png)

**Comment by amrdraz** [comment]
@skidding works great for the leaf node of the tree, but doesn't seem to work for components in the middle ie. a component with fixtures defined and has another folder with other fixtures in them.
only the final component is shown as a single folder

<img width="236" alt="screen shot 2017-10-21 at 23 06 56" src="https://user-images.githubusercontent.com/1708559/31855884-c41b9f7c-b6b4-11e7-83ac-38e5527a37a4.png">

You will notice `OptionTitle` is not repeated but `OptionsHeader` and `OptionsList` are because they have nested components

<img width="332" alt="screen shot 2017-10-21 at 23 06 00" src="https://user-images.githubusercontent.com/1708559/31855909-2df8b902-b6b5-11e7-862f-dd8405ad8e3f.png">

I realised that NavigationTopBar was showing the way it was because that's what I called the component, this is also why in the screenshot `OptionsHeader` is repeated as `OptionHeader` which I am realizing is a typo in the class name, this is not an issue, if anything it's nice to notice 😅 .

**Comment by skidding** [comment]
@amrdraz I'm still open to alternatives, but did you have a chance the latest version (3.0.0-alpha.10) with the component icons? It might seem more natural since folders are distinguished from components.

**Comment by amrdraz** [comment]
@skidding nice!! 

it looks really good in distinguishing components inside folders
<img width="335" alt="screen shot 2017-10-24 at 14 26 30" src="https://user-images.githubusercontent.com/1708559/31942971-70ea5b3a-b8c7-11e7-8463-a742e129b463.png">

this is how my nested monster looks like now

<img width="322" alt="screen shot 2017-10-24 at 14 19 49" src="https://user-images.githubusercontent.com/1708559/31942905-372c63d4-b8c7-11e7-8e9a-b051d3d685b8.png">

at first I wasn't so sure about the case with nested components not showing up as components

<img width="330" alt="screen shot 2017-10-24 at 14 31 27" src="https://user-images.githubusercontent.com/1708559/31943189-371b83ce-b8c8-11e7-8e8f-84796b705a3b.png">

but after thinking about it I guess it creates a kind of distinction
I would still prefer if a component shows up as a component even if it has nested content

But I think those that don't distinguish between nested components with a `components` folder would start having a tree of components which may render the point of having icons useless.

then again it would be an accurate representation of their system 🤔 

well I'm satisfied with this so I think this is great 👍

**Comment by skidding** [comment]
> it looks really good in distinguishing components inside folders

Indeed, and we have @maxsalven to thank!

> this is how my nested monster looks like now

In the future we could have an option to hook into the function that generates the namespace based on file path. In your case you could apply a transformation to trim `components/` (I think you mentioned this in the past) and it would make the tree even nicer.

> But I think those that don't distinguish between nested components with a components folder would start having a tree of components which may render the point of having icons useless.

I'm sure we'll iterate on this in the future, but for know I like the distinction:
- If you see a component icon you can expect fixtures inside
- If you see a folder icon you can expect components and/or folders inside

> well I'm satisfied with this so I think this is great 👍

👍

**Comment by skidding** [comment]
Please subscribe to [Cosmos v3 BETA TESTING thread](https://github.com/react-cosmos/react-cosmos/issues/475). Conversation will continue there.
